src/middlewared/middlewared/plugins/etc.py

```diff
@@ -335,12 +339,25 @@ def __init__(self, *args, **kwargs):
             'py': PyRenderer(self),
         }
 
+    async def gather_context(self, methods):
+        rv = {}
+        for m in methods:
+            rv[m] = await self.middleware.call(m)
+
+        return rv
+
     async def generate(self, name, checkpoint=None):
         group = self.GROUPS.get(name)
         if group is None:
             raise ValueError('{0} group not found'.format(name))
 
         async with self.LOCKS[name]:
+            if group[0]['type'] == 'ctx':
```

Comment:
Looks a little bit hacky. Can we just change definition style for groups that require a context? Something like
```
        'nfsd': {
            'ctx': {'methods': ['sharing.nfs.query', 'nfs.config']}},
            'entries': [
                {'type': 'mako', 'path': 'default/nfs-common'},
                {'type': 'mako', 'path': 'default/nfs-kernel-server'},
            ],
```

Reply:
Good idea.